types/basetypes/tuple_value.go

```diff
@@ -0,0 +1,254 @@
+				fmt.Sprintf("Tuple Expected Type: %v\n", elementTypes)+
+				fmt.Sprintf("Tuple Given Type: %v", givenTypes),
+		)
+
+		return NewTupleUnknown(elementTypes), diags
+	}
+
+	for i, element := range elements {
+		if !elementTypes[i].Equal(element.Type(ctx)) {
+			diags.AddError(
+				"Invalid Tuple Element",
+				"While creating a Tuple value, an invalid element was detected. "+
+					"A Tuple must be an ordered array of elements where the values exactly match the length and types of the defined element types. "+
+					"This is always an issue with the provider and should be reported to the provider developers.\n\n"+
+					fmt.Sprintf("Tuple Index (%d) Expected Type: %s\n", i, elementTypes[i])+
+					fmt.Sprintf("Tuple Index (%d) Given Type: %s", i, element.Type(ctx)),
+			)
+		}
+	}
+
+	if diags.HasError() {
+		return NewTupleUnknown(elementTypes), diags
+	}
+
+	return TupleValue{
+		elementTypes: elementTypes,
+		elements:     elements,
+		state:        attr.ValueStateKnown,
+	}, nil
+}
+
+// NewTupleValueMust creates a Tuple with a known value, converting any diagnostics
+// into a panic at runtime. Access the value via the Tuple type Elements method.
+//
+// This creation function is only recommended to create Tuple values which will
+// not potentially affect practitioners, such as testing, or exhaustively
+// tested provider logic.
+func NewTupleValueMust(elementTypes []attr.Type, elements []attr.Value) TupleValue {
+	tuple, diags := NewTupleValue(elementTypes, elements)
+
+	if diags.HasError() {
+		// This could potentially be added to the diag package.
+		diagsStrings := make([]string, 0, len(diags))
+
+		for _, diagnostic := range diags {
+			diagsStrings = append(diagsStrings, fmt.Sprintf(
+				"%s | %s | %s",
+				diagnostic.Severity(),
+				diagnostic.Summary(),
+				diagnostic.Detail()))
+		}
+
+		panic("NewTupleValueMust received error(s): " + strings.Join(diagsStrings, "\n"))
+	}
+
+	return tuple
+}
+
+// TupleValue represents an ordered list of attr.Value, with an attr.Type for each element. This type intentionally
+// includes less functionality than other types in the type system as it has limited real world application and therefore
+// is not exposed to provider developers.
+type TupleValue struct {
+	// elements is the ordered list of known element values for the tuple.
+	elements []attr.Value
+
+	// elementTypes is the ordered list of elements types for the tuple.
+	elementTypes []attr.Type
+
+	// state represents whether the value is null, unknown, or known. The
+	// zero-value is null.
+	state attr.ValueState
+}
+
+// Elements returns a copy of the ordered list of known values for the Tuple.
+func (v TupleValue) Elements() []attr.Value {
+	// Ensure callers cannot mutate the internal elements
+	result := make([]attr.Value, 0, len(v.elements))
+	result = append(result, v.elements...)
+
+	return result
+}
+
+// ElementTypes returns a copy of the ordered list of element types for the Tuple.
```

Comment:
To match the current implementation.

```suggestion
// ElementTypes returns the ordered list of element types for the Tuple.
```